tests/test_federation.py

```diff
@@ -276,3 +276,111 @@ def test_cross_signing_keys_retry(self):
+    def setUp(self):
+        super().setUp()
+
+    def strip_unauthorized_unsigned_values(self):
+        most_recent = self.get_success(
+            self.homeserver.get_datastore().get_latest_event_ids_in_room(self.room_id)
+        )[0]
+        federation_event_handler = self.homeserver.get_federation_event_handler()
+        event1 = make_event_from_dict(
+            {
+                "room_id": self.room_id,
+                "sender": "@baduser:test.serv",
+                "state_key": "@baduser:test.serv",
+                "event_id": "$event1:test.serv",
+                "depth": 1000,
+                "origin_server_ts": 1,
+                "type": "m.room.member",
+                "origin": "test.servx",
+                "content": {"membership": "join"},
+                "auth_events": [],
+                "prev_state": [(most_recent, {})],
+                "prev_events": [(most_recent, {})],
+                "unsigned": {"malicious garbage": "hackz", "more warez": "more hackz"},
+            }
+        )
+        self.get_success(federation_event_handler.on_receive_pdu("test.serv", event1))
+
+        event = self.get_success(self.store.get_event("$event1:test.serv"))
+        event_dict = event.get_dict()
+        # Make sure unauthorized fields are stripped from unsigned
+        self.assertNotIn("hackz", event_dict["unsigned"])
+
+    def strip_event_maintains_allowed_fields(self):
+        most_recent = self.get_success(
+            self.homeserver.get_datastore().get_latest_event_ids_in_room(self.room_id)
+        )[0]
+        federation_event_handler = self.homeserver.get_federation_event_handler()
+        event2 = make_event_from_dict(
+            {
+                "room_id": self.room_id,
+                "sender": "@baduser:test.serv",
+                "state_key": "@baduser:test.serv",
+                "event_id": "$event2:test.serv",
+                "depth": 1000,
+                "origin_server_ts": 1,
+                "type": "m.room.member",
+                "origin": "test.servx",
+                "auth_events": [],
+                "prev_state": [(most_recent, {})],
+                "prev_events": [(most_recent, {})],
+                "content": {"membership": "join"},
+                "unsigned": {
+                    "malicious garbage": "hackz",
+                    "more warez": "more hackz",
+                    "age": 14,
+                    "invite_room_state": [],
+                },
+            }
+        )
+        self.get_success(
+            federation_event_handler.on_send_membership_event("test.serv", event2)
+        )
+
+        event = self.get_success(self.store.get_event("$event2:test.serv"))
+        event_dict = event.get_dict()
+        self.assertIn("age", event_dict["unsigned"])
+        self.assertNotIn("hackz", event_dict["unsigned"])
+        # Invite_room_state is allowed in events of type m.room.member
+        self.assertIn("invite_room_state", event_dict["unsigned"])
```

Comment:
Could also check that the values for `age` and `invite_room_state` are unaltered by your change.